File: src/main/java/com/github/simiacryptus/aicoder/actions/MarkdownNewTableRowsAction.java
package com.github.simiacryptus.aicoder.actions;

import com.github.simiacryptus.aicoder.config.AppSettingsState;
import com.github.simiacryptus.aicoder.util.StringTools;
import com.github.simiacryptus.aicoder.util.UITools;
import com.github.simiacryptus.aicoder.util.psi.PsiUtil;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Document;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static com.github.simiacryptus.aicoder.util.UITools.insertString;

public class MarkdownNewTableRowsAction extends AnAction {

    @Override
    public void update(@NotNull AnActionEvent e) {
        e.getPresentation().setEnabledAndVisible(isEnabled(e));
        super.update(e);
    }

    private static boolean isEnabled(@NotNull AnActionEvent e) {
        return null != getMarkdownNewTableRowsParams(e);
    }

    @Nullable
    public static MarkdownNewTableRowsParams getMarkdownNewTableRowsParams(@NotNull AnActionEvent e) {
        Caret caret = e.getData(CommonDataKeys.CARET);
        if (null != caret) {
            PsiFile psiFile = e.getData(CommonDataKeys.PSI_FILE);
            if (null != psiFile) {
                PsiElement table = PsiUtil.getSmallestIntersecting(psiFile, caret.getSelectionStart(), caret.getSelectionEnd(), "MarkdownTableImpl");
                if (null != table) {
                    return new MarkdownNewTableRowsParams(caret, table);
                }
            }
        }
        return null;
    }

    @Override
    public void actionPerformed(@NotNull AnActionEvent event) {
        MarkdownNewTableRowsParams markdownNewTableRowsParams = getMarkdownNewTableRowsParams(event);
        List<CharSequence> rows = StringTools.trim(PsiUtil.getAll(Objects.requireNonNull(markdownNewTableRowsParams).table, "MarkdownTableRowImpl")
                .stream().map(PsiElement::getText).collect(Collectors.toList()), 10, true);
        CharSequence n = Integer.toString(rows.size() * 2);
        AppSettingsState settings = AppSettingsState.getInstance();
        int endOffset = markdownNewTableRowsParams.table.getTextRange().getEndOffset();
        Document document = event.getRequiredData(CommonDataKeys.EDITOR).getDocument();
        UITools.redoableRequest(MarkdownNewTableColAction.newRowsRequest(settings, n, rows, ""), "", event,
                newText -> transformCompletion(markdownNewTableRowsParams, newText),
                newText -> insertString(document, endOffset, newText));
    }

    @NotNull
    private static String transformCompletion(MarkdownNewTableRowsParams markdownNewTableRowsParams, CharSequence complete) {
        CharSequence indent = UITools.getIndent(markdownNewTableRowsParams.caret);
        List<CharSequence> newRows = Arrays.stream(("" + complete).split("\n"))
                .map(String::trim).filter(x -> x.length() > 0).collect(Collectors.toList());
        return "\n" + indent + newRows.stream().collect(Collectors.joining("\n" + indent));
    }

    public static class MarkdownNewTableRowsParams {
        public final Caret caret;
        public final PsiElement table;

        private MarkdownNewTableRowsParams(Caret caret, PsiElement table) {
            this.caret = caret;
            this.table = table;
        }

    }
}

File: src/main/kotlin/com/github/simiacryptus/aicoder/actions/markdown/MarkdownNewTableRowsAction.kt
package com.github.simiacryptus.aicoder.actions.markdown

import com.github.simiacryptus.aicoder.config.AppSettingsState
import com.github.simiacryptus.aicoder.util.ComputerLanguage
import com.github.simiacryptus.aicoder.util.StringTools
import com.github.simiacryptus.aicoder.util.UITools
import com.github.simiacryptus.aicoder.util.psi.PsiUtil
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.editor.Caret
import com.intellij.psi.PsiElement
import java.util.*
import java.util.stream.Collectors

class MarkdownNewTableRowsAction : AnAction() {
    override fun update(e: AnActionEvent) {
        e.presentation.isEnabledAndVisible = isEnabled(e)
        super.update(e)
    }

    override fun actionPerformed(event: AnActionEvent) {
        val markdownNewTableRowsParams = getMarkdownNewTableRowsParams(event)
        val rows = StringTools.trim(
            PsiUtil.getAll(
                markdownNewTableRowsParams!!.table, "MarkdownTableRowImpl"
            ).stream().map { obj: PsiElement -> obj.text!! }.collect(Collectors.toList()), 10, true
        )
        val n: CharSequence = Integer.toString(rows.size * 2)
        val settings = AppSettingsState.instance
        val endOffset = markdownNewTableRowsParams.table.textRange.endOffset
        val document = event.getRequiredData(CommonDataKeys.EDITOR).document
        UITools.redoableRequest(
            MarkdownNewTableColAction.newRowsRequest(settings, n, rows, ""),
            "",
            event,
            { transformCompletion(markdownNewTableRowsParams, it) },
            { UITools.insertString(document, endOffset, it) })
    }

    class MarkdownNewTableRowsParams constructor(val caret: Caret, val table: PsiElement)
    companion object {
        private fun isEnabled(e: AnActionEvent): Boolean {
            if (UITools.isSanctioned()) return false
            val computerLanguage = ComputerLanguage.getComputerLanguage(e) ?: return false
            return if (ComputerLanguage.Markdown !== computerLanguage) false else null != getMarkdownNewTableRowsParams(
                e
            )
        }

        fun getMarkdownNewTableRowsParams(e: AnActionEvent): MarkdownNewTableRowsParams? {
            val caret = e.getData(CommonDataKeys.CARET)
            if (null != caret) {
                val psiFile = e.getData(CommonDataKeys.PSI_FILE)
                if (null != psiFile) {
                    val table =
                        PsiUtil.getSmallestIntersecting(
                            psiFile,
                            caret.selectionStart,
                            caret.selectionEnd,
                            "MarkdownTableImpl"
                        )
                    if (null != table) {
                        return MarkdownNewTableRowsParams(caret, table)
                    }
                }
            }
            return null
        }

        private fun transformCompletion(
            markdownNewTableRowsParams: MarkdownNewTableRowsParams,
            complete: CharSequence?
        ): String {
            val indent = UITools.getIndent(markdownNewTableRowsParams.caret)
            val newRows = Arrays.stream(("" + complete).split("\n".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray())
                .map { obj: String -> obj.trim { it <= ' ' } }.filter { x: String -> x.length > 0 }
                .collect(Collectors.toList())
            return "\n$indent" + newRows.stream().collect(Collectors.joining("\n$indent"))
        }
    }
}